test/trace/test_trace.py

```diff
@@ -0,0 +1,132 @@
+import pytest
+from hpy.trace import call_counts, durations, set_trace_functions
+
+@pytest.fixture
+def hpy_abi():
+    return "trace"
+
+def _relative_call_count(name, call_count_0, diff):
+    return call_counts()[name] == call_count_0[name] + diff
+
+
+def _assert_unchanged_except(expected, actual, *ignore):
+    import pytest
+    for x in expected.keys():
+        if x not in ignore and actual[x] != expected[x]:
+            pytest.fail("Call count of %s unexpectedly changed (expected %d but was %d)." %
+                        (x, expected[x], call_counts()[x]))
+
+
+def test_call_counts(compiler):
+    import pytest
+    mod = compiler.make_module("""
+        HPyDef_METH(f, "f", f_impl, HPyFunc_VARARGS)
+        static HPy f_impl(HPyContext *ctx, HPy self, HPy *args, HPy_ssize_t nargs)
+        {
+            if (nargs != 2) {
+                HPyErr_SetString(ctx, ctx->h_TypeError, "expected exactly two args");
+                return HPy_NULL;
+            }
+            return HPy_Add(ctx, args[0], args[1]);
+        }
+
+        @EXPORT(f)
+        @INIT
+    """)
+    # Don't rely on absolute call count numbers since the module setup could
+    # already do some API calls we don't expect.
+    call_counts_0 = call_counts().copy()
+
+    assert mod.f(1, 2) == 3
+    assert _relative_call_count("ctx_Add", call_counts_0, 1)
+    _assert_unchanged_except(call_counts_0, call_counts(), "ctx_Add")
```

Comment:
again, if you use `collections.Counter` you can greatly simplify your code here. E.g.:
```
>>> import collections
>>> a = collections.Counter({'dup': 3, 'close': 4})
>>> b = collections.Counter({'dup': 3, 'close': 5})
>>> b -a
Counter({'close': 1})
```

So you could say something like this:
```python
counts = call_counts() - call_counts_0
assert counts == {'ctx_Add': 1}
```

much better :)